lib/web_of_science/process_records.rb

```diff
@@ -0,0 +1,134 @@
+module WebOfScience
+
+  # This class complements the WebOfScience::Harvester
+  # Process records retrieved by any means; this is a progressive filtering of the harvested records to identify
+  # those records that should create a new Publication-PubHash, PublicationIdentifier(s) and Contribution(s).
+  class ProcessRecords
+
+    # @param author [Author]
+    # @param records [WebOfScience::Records]
+    def initialize(author, records)
+      raise(ArgumentError, 'author must be an Author') unless author.is_a? Author
+      raise(ArgumentError, 'records must be an WebOfScience::Records') unless records.is_a? WebOfScience::Records
+      @author = author
+      @records = records
+    end
+
+    # @return [Array<String>] WosUIDs that create a new Publication
+    def execute
+      return [] if records.count.zero?
+      create_publications
+      # TODO: batch harvest PubMed data for WOS records with a PMID?
+    rescue StandardError => err
+      logger.error(err.inspect)
+      []
+    end
+
+    private
+
+      attr_reader :author
+      attr_reader :records
+
+      # ----
+      # Record filters and data flow steps
+
+      # @return [Array<String>] WosUIDs that create a new Publication
+      def create_publications
+        new_wos_records = match_wos_records # cf. WebOfScienceSourceRecord
+        saved_wos_records = save_wos_records(new_wos_records) # save WebOfScienceSourceRecord
+        records = filter_by_identifiers(saved_wos_records) # cf. PublicationIdentifier
+        records.map { |rec| create_publication(rec) }.compact
+      end
+
+      ## 1
+      # Filter and select new WebOfScienceSourceRecords
+      # @return [Array<WebOfScience::Record>]
+      def match_wos_records
+        return [] if records.count.zero?
+        records.select { |rec| WebOfScienceSourceRecord.where(uid: rec.uid).count.zero? }
+      end
+
+      ## 2
+      # Save and select new WebOfScienceSourceRecords
+      # @param [Array<WebOfScience::Record>]
+      # @return [Array<WebOfScience::Record>]
+      def save_wos_records(records)
+        # IMPORTANT: add nothing to PublicationIdentifiers here, or new_records will reject them
+        return [] if records.empty?
+        # We only want the 'pmid' for "WOS" records ("MEDLINE" records have one already)
+        process_links(records.select { |rec| rec.database == 'WOS' })
+        records.select do |rec|
+          saved = WebOfScienceSourceRecord.new(source_data: rec.to_xml).save!
```

Comment:
isn't it enough to do something like this instead of using the intermediate `saved` variable?  `records.select {|rec| WebOfScienceSourceRecord.new(source_data: rec.to_xml).save! }`

Reply:
Thanks, aware of this.  This section will be changed to add identifiers to the src-record, but that TODO item depends on doing the work to add a migration to the src-record to add the fields for the identifiers.  Unlike other src-models, this one needs outside assistance to save some of the identifiers from the links-API.  I'm aware of this and left the intermediate variable and TODO note there intentionally as a reminder and sign-post that this is where a pending revision will occur in the near sprints.